File: 04_sorting/04_06_bubble_sort_recursive/bubble_sort_recursive.cpp
#include <bits/stdc++.h>

using namespace std;

void printArray(int arr[], int n)
{
    for (size_t i = 0; i < n; i++)
    {
        cout << arr[i] << " ";
    }
}

void bubble_sort_recursive(int arr[], int n)
{
    if (n == 1)
    {
        return;
    }
    
    for (size_t i = 0; i < n - 1; i++)
    {
        if (arr[i] > arr[i+1])
        {
            swap(arr[i], arr[i+1]);
        }
    }
    
    bubble_sort_recursive(arr, n-1);
}

int main()
{
    int arr[] = {13, 46, 24, 52, 20, 9};
    int n = sizeof(arr) / sizeof(arr[0]);

    cout << "Before:-" << endl;
    printArray(arr, n);
    bubble_sort_recursive(arr, n);
    cout << endl;
    cout << "After:-" << endl;
    printArray(arr, n);
}

File: 04_sorting/04_06_bubble_sort_recursive/BubbleSortRecursive.java
public class BubbleSortRecursive {
    public static void main(String[] args) {
        int arr[] = { 13, 46, 24, 52, 20, 9 };
        System.out.println("Before:-");
        print(arr);
        bubbleSortRecursive(arr, arr.length);
        System.out.println();
        System.out.println("After:-");
        print(arr);
    }

    private static void bubbleSortRecursive(int arr[], int n) {
        if (n == 1) {
            return;
        }

        for (int i = 0; i < n - 1; i++) {
            if (arr[i] > arr[i + 1]) {
                swap(arr, i+1, i);
            }
        }

        bubbleSortRecursive(arr, n-1);
    }

    private static void swap(int[] arr, int a, int b) {
        int temp = arr[a];
        arr[a] = arr[b];
        arr[b] = temp;
    }

    private static void print(int[] arr) {
        for (int i = 0; i < arr.length; i++) {
            System.out.print(arr[i] + " ");
        }
    }
}
